### DISCUSSION TITLE
Use w/ 3rd party logging crate

### DISCUSSION THREAD

**Discussion by kpeters-cbsi** [tokio-rs/tracing#1830] [Q&A] [answered]
I'm working on a Rust project using Fastly's [Compute@Edge](https://docs.fastly.com/products/compute-at-edge) platform. C@E allows you to choose a variety of logging backends; I'm using New Relic, which takes JSON input. The nice thing is that, if you use Fastly's [log_fastly crate](https://docs.rs/log-fastly/latest/log_fastly/) Fastly will handle your New Relic authentication for you, so all you do is `log::info!(...)` and you're good. The not-so-nice thing is that while the crate will automagically format a simple string as JSON for you (in the form `{"message":"whatever you put in your simple string"}`) it doesn't support adding fields to every message, so you have to do something like `log::info!(r#"{"message":"something", "source":"myapp"}"#);` every time. This is a pain in the ass. Is there a way to get `tracing` to call Fastly's `log!()` macro?

**Comment by hawkw** [comment, marked as answer]
`log_fastly` appears to re-export the `log` crate's macros. If you want to use `tracing` to export to the `log` crate, you can use the ["log" feature flag](https://docs.rs/tracing/latest/tracing/#emitting-log-records) on the `tracing` crate to send `tracing` events to the `log` crate. However, this _won't_ actually get you structured fields reported as JSON; they will just be formatted as `key=value` in a string.

If you want to actually use `tracing` to emit structured JSON logs, you would need to actually use a `tracing::Subscriber` implementation that consumes `tracing`'s structured fields. Unfortunately, I don't think that will work nicely out of the box with `log_fastly`. If you want to continue using Fastly's logging infrastructure, rather than setting up your own `tracing` subscriber, the best solution is _probably_ to ask the Fastly team if they'd consider adding `tracing` support to `log_fastly`. I know some of the Compute@Edge team (hi @acfoltzer, @pchickey,  and @cratelyn!) are `tracing` enthusiasts, so it's possible they could add a `tracing::Subscriber` implementation in `log_fastly`!

**Reply by acfoltzer** [reply]
Hi @hawkw! Yes indeed, a tracing subscriber backed by the lower-level [`fastly::log::Endpoint`](https://docs.rs/fastly/latest/fastly/log/struct.Endpoint.html) type is what's needed here. It should be possible in the mean time to implement without any internal changes to the existing `fastly` crate, but I've made sure to pass this along as a potential TODO to provide a first-party solution.

**Reply by hawkw** [reply]
I think all that's necessary is to add a [`MakeWriter`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/writer/trait.MakeWriter.html) implementation for `fastly::log::Endpoint` (or a wrapper type, in downstream code), and then it could be passed to `tracing_subscriber::fmt`'s [`SubscriberBuilder::with_writer`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.SubscriberBuilder.html#method.with_writer) method. It would then be necessary to configure the subscriber to format events as JSON.

Eventually, fastly could probably provide a wrapper that configures `tracing` to do that, so that it just works out of the box, similarly to the `log` crate.

**Reply by kpeters-cbsi** [reply]
So... how would I go about doing that? I'm still pretty new to Rust.

**Reply by kpeters-cbsi** [reply]
Well, I couldn't figure out the `MakeWriter` business, but I did find a lovely blog post at https://burgers.io/custom-logging-in-rust-using-tracing that helped a great deal. I took the `CustomLayer` [created there](https://github.com/bryanburgers/tracing-blog-post/blob/main/examples/figure_10/custom_layer.rs) and in place of `println!("{}", serde_json::to_string_pretty(&output).unwrap());`, I did `writeln!(endpoint, "{}", serde_json::to_string(&output).unwrap()).unwrap();` (`endpoint` being an instance of `Fastly::log::Endpoint`), and that looks to have solved my problem.
